tao_compiler/mlir/disc/transforms/lower_tf.cc

```diff
@@ -660,10 +661,33 @@ Type ToLegalElementType(Type type) {
       .Default([&type](Type) { return type; });
 }
 
+bool isQint32ConstantZeroTensor(Value v) {
+  auto constOp = dyn_cast_or_null<TF::ConstOp>(v.getDefiningOp());
+  if (!constOp) return false;
+
+  tensorflow::Tensor out;
+  if (tensorflow::ConvertToTensor(constOp.value(), &out) !=
+      tensorflow::Status::OK())
+    return false;
+
+  int64_t numElems = out.NumElements();
+  auto flat = out.bit_casted_shaped<int32_t, 1>({numElems});
+  for (int64_t i = 0; i < numElems; ++i) {
+    if (flat(i) != 0) return false;
+  }
+  // in case the value is an empty tensor.
+  return (numElems != 0);
+}
+
 bool isConstantZeroTensor(Value v) {
   auto castOp = dyn_cast_or_null<TF::CastOp>(v.getDefiningOp());
   if (castOp) return isConstantZeroTensor(castOp->getOperand(0));
 
+  auto ty = v.getType().dyn_cast<RankedTensorType>();
+  if (ty && ty.getElementType().isa<mlir::TF::Qint32Type>()) {
+    return isQint32ConstantZeroTensor(v);
```

Comment:
why Qint32 is allowed to to have multiple value in tensor?

Reply:
The `i32` or `float` case also supports zero tensors with `num elements > 1`. Here zero tensors are those contents are all zeros, no matter how many elements it has.  

The different between the `f32/i32` case and `qint32` case is:
- The content of a qint32 tensor is a tensorflow tensor object, which does not support `splat` storage.
- The content of a i32/f32 tensor is a mlir dense attributes, which supports `splat` storage. If all the elements of a tensor are zeros, then it could use splat storage, in which case we only need to check the first elements.